File: Mesh3D/Matlab3D/examples/example_bunny.m
clear;
clc;
close all;

projectRoot = fileparts(fileparts(mfilename('fullpath')));
addpath(projectRoot);
addpath(fullfile(projectRoot, 'src'));

%**************************************************************************
%
%   Esempio: 
%       Cubatura su dominio generico
%
%   Descrizione:
%       Questo esempio dimostra l'utilizzo del metodo OptimalPolyCuba3D per
%       l'integrazione di una funzione su un dominio poliedrale generico,
%       rappresentato da una mesh superficiale triangolare. 
%
%**************************************************************************

%**************************************************************************
%  Parametri e caricamento mesh
%**************************************************************************

ade = 133;

fprintf('\n');
fprintf('**************************************************************\n');
fprintf('                 OPTIMALPOLYCUBA3D\n');
fprintf('        Cubatura su Dominio Poliedrale "Stanford Bunny" \n');
fprintf('**************************************************************\n');
fprintf('\n');

fprintf('Ade:                   %d\n', ade);

mesh = readSurfaceMesh('bun_zipper.ply');

vertices = mesh.Vertices;
facets   = mesh.Faces;

fprintf('Numero di vertici:     %d\n', size(vertices,1));
fprintf('Numero di facce:       %d\n', size(facets,1));

%**************************************************************************
% Definizione funzione integranda
%**************************************************************************

f = @(x,y,z) ones(size(x));
f_string = 'f(x,y,z) = 1';
fprintf('Funzione integranda:   %s\n', f_string);

%**************************************************************************
% Plot dominio poliedrale
%**************************************************************************

% Riduzione della mesh esclusivamente per la visualizzazione
% 1% dei triangoli utilizzati nella cubatura
[facets_plot, vertices_plot] = reducepatch(facets, vertices, 0.01);

% Rotazione della mesh solo per la visualizzazione
vertices_plot = vertices_plot * [1  0  0;
                                  0  0  1;
                                  0 -1  0];

figure('Color','w');

patch('Vertices', vertices_plot, ...
      'Faces', facets_plot, ...
      'FaceColor', [0.75 0.75 0.75], ...
      'EdgeColor', [0.15 0.15 0.15], ...
      'LineWidth', 0.5);

axis equal;
grid on;
box on;
view(3);

xlabel('$x$', 'Interpreter','latex');
ylabel('$y$', 'Interpreter','latex');
zlabel('$z$', 'Interpreter','latex');

title('Stanford Bunny', 'Interpreter','latex');

%**************************************************************************
% Inizio regola di cubatura
%**************************************************************************

fprintf('\n');
fprintf('Inizio Cubatura...\n');

tic;
[XYZ, W] = OPC3D(ade, vertices, facets, "D");
%[XYZ, W] = OPC3D(ade, vertices, facets, "GJ");
elapsedTime = toc;

I = W' * f(XYZ(:,1), XYZ(:,2), XYZ(:,3));

%**************************************************************************
% Visualizzazione risultati e punti di cubatura
%**************************************************************************

fprintf('Fine Cubatura...\n');

fprintf('\n');
fprintf('Integrale numerico:    %.15e\n', I);
fprintf('Tempo di calcolo:      %.6e s\n', elapsedTime);

File: Mesh3D/Fortran3D/examples/example_bunny.f90
PROGRAM example_bunny

USE TypesDef, ONLY: dp
USE PolyhedronMesh, ONLY: t_polyhedron, MeshReader
USE OPC3D_Module, ONLY: OPC3D

    IMPLICIT NONE

    !***********************************************************************
    !
    !   Esempio:
    !       Cubatura su dominio poliedrale generico
    !
    !   Descrizione:
    !       Questo esempio dimostra l'utilizzo del metodo
    !       OptimalPolyCuba3D per l'integrazione di una funzione su un
    !       dominio poliedrale generico, rappresentato da una mesh
    !       superficiale triangolare.
    !
    !***********************************************************************

    !***********************************************************************
    ! Dichiarazione delle variabili
    !***********************************************************************

    INTEGER                         :: ade
    INTEGER                         :: n_vertici, n_facce
    REAL(dp)                        :: Integrale
    INTEGER                         :: i
    REAL(dp), ALLOCATABLE           :: XYZ(:,:)
    REAL(dp), ALLOCATABLE           :: W(:)
    REAL(dp)                        :: elapsedTime
    REAL(dp)                        :: t_start, t_end

    TYPE(t_polyhedron)              :: poly

    CHARACTER(LEN=256)              :: vertici_file
    CHARACTER(LEN=256)              :: tri_file

    !***********************************************************************
    ! Parametri e caricamento mesh
    !***********************************************************************

    ade = 2

    WRITE(*,'(A)')
    WRITE(*,'(A)') '**************************************************************'
    WRITE(*,'(A)') '                            OPC3D'
    WRITE(*,'(A)') '        Cubatura su Dominio Poliedrale "Stanford Bunny"'
    WRITE(*,'(A)') '**************************************************************'
    WRITE(*,'(A)')

    WRITE(*,'(A,I0)') 'Ade:                   ', ade

    ! Percorsi relativi alla cartella examples.
    vertici_file = 'bunny_vertex.dat'
    tri_file     = 'bunny_tri.dat'

    ! Lettura della mesh superficiale triangolare.
    CALL MeshReader(vertici_file, tri_file, poly)

    n_vertici = SIZE(poly%vertici,1)
    n_facce   = SIZE(poly%facce,1)

    WRITE(*,'(A,I0)') 'Numero di vertici:     ', n_vertici
    WRITE(*,'(A,I0)') 'Numero di facce:       ', n_facce

    !***********************************************************************
    !   Definizione funzione integranda
    !***********************************************************************

    WRITE(*,'(A)') 'Funzione integranda:   f(x,y,z) = x^2+y^2+z^2'

    !***********************************************************************
    !   Inizio regola di cubatura
    !***********************************************************************

    WRITE(*,'(A)')
    WRITE(*,'(A)') 'Inizio Cubatura...'

    CALL CPU_TIME(t_start)

    CALL OPC3D(ade, poly%vertici, poly%facce, 'D', XYZ, W)
    ! CALL OPC3D(ade, poly%vertici, poly%facce, 'D', XYZ, W)

    CALL CPU_TIME(t_end)

    elapsedTime = t_end - t_start

    Integrale = 0.0_dp
    DO i = 1, SIZE(W)
        Integrale = Integrale + W(i) * f(XYZ(i,1), XYZ(i,2), XYZ(i,3))
    END DO

    !***********************************************************************
    ! Visualizzazione risultati e punti di cubatura
    !***********************************************************************

    WRITE(*,'(A)') 'Fine Cubatura...'

    WRITE(*,'(A)')
    WRITE(*,'(A,F0.15)') 'Integrale numerico:       ', Integrale
    WRITE(*,'(A,ES14.6,A)') 'Tempo di calcolo:       ', elapsedTime, ' s'

    CONTAINS

    !***********************************************************************
    ! Funzione integranda
    !***********************************************************************

    REAL(dp) FUNCTION f(x,y,z)

        IMPLICIT NONE

        REAL(dp), INTENT(IN) :: x
        REAL(dp), INTENT(IN) :: y
        REAL(dp), INTENT(IN) :: z

        f = x**2+y**2+z**2

    END FUNCTION f

END PROGRAM example_bunny